### PR TITLE
[cmd][diff] Ignore detection status for tags

### PR DESCRIPTION
**Pull request by Szelethus**
We already do this on the GUI, but on the cmdline, we default the
detection status to NEW, REOPENED and UNRESOLVED. This makes sense for
runs (where we check whether a report is outstanding at the time of the
query), but not for tags (as a report can have a RESOLVED detection
status but still be outstanding at the time of the tag).

### CODE REVIEW

**Review by bruntib — APPROVED** [review]
Please, rebase. Otherwise looks good.

File: docs/web/diff.md [hunk 1]
```diff
@@ -0,0 +1,902 @@
+# Comparing analysis results (Diff)
+
+Compare the outstanding reports in a "before" (_baseline_) and an "after" (_new_ or newline) analysis sets, and display the differences in between them the in the same
```
Comment (dkrupp) [inline docs/web/diff.md:3]: Please start this section with a more descriptive introduction of the report states:

I would rewrite the starting stentence like this:

In CodeChecker a report can be either "Outstanding", or "Closed". Outstanding reports are the reports which should be fixed: are still detected, are unreviewed or confirmed to be true positive.
We call a reports "Closed", if it cannot be detected anymore (e.g. because it is fixed), or it is considered a false positive or an uninteresting (intentional) finding.

CodeChecker diff mechanism always compares two outstanding set of reports, a "before" (_baseline_) and an "after" (_new_ or newline) analysis sets, and display the differences in between them.

Formally:

A report is outsstanding if all ....

File: docs/web/diff.md [hunk 2]
```diff
@@ -0,0 +1,902 @@
...
+  ```
+  ----======== Summary ========----
+  ---------------------------------------------
+  Number of processed analyzer result files | 0
+  Number of analyzer reports                | 0
+  ---------------------------------------------
+  ----=================----
+  ```
+
+## Review status rules and diffs
```
Comment (dkrupp) [inline docs/web/diff.md:611]: Refer to the use cases introduced in the introduction section.

E.g.

If you whish to store the suppressions on the server, you can use review status rules. With the review status rule you can set all instances of report type (reports with the same hash) e.g. false positive. This feature makes it easy to mark a report false positive in all runs at the same time.

File: docs/web/diff.md [hunk 3]
```diff
@@ -0,0 +1,902 @@
...
+  ---------------------------------------------
+  Number of processed analyzer result files | 0
+  Number of analyzer reports                | 0
+  ---------------------------------------------
+  ----=================----
+  ```
+
+## Diff on tags or timestamps
+
+You can filter the analysis results or diff results based on which reports were outstanding _before_ a given timestamp. In the below image, on the web GUI, we diff the reports outstanding before the 1st of January, 2023 in "Test Run" against reports  outstanding before the 1st of January, 2023 in "Test Run Updated":
```
Comment (dkrupp) [inline docs/web/diff.md:751]: Refer to the introduction section "server based branch history analysis".

If a run is continously updated with new analysis results, it can store the past report states of all previous analysis executions. This can be useful to analyze how the reports were fixed, classified to false positives, or introduced over the development timeline.

File: docs/web/diff.md [hunk 4]
```diff
@@ -0,0 +1,902 @@
...
+
+:warning: Note: Tags are created when the report is stored. As a result, you can diff a local analysis directory against a remote tag, but there are no "local tags".
+
+:no_entry: Note: You can compare tags through the command line and the web GUI as well. Currently, you can only compare timestamps on the web GUI.
+
+:no_entry: Note: Comparing tags older than the second-to-last is not supported, and may yield incorrect results in specific circumstances. Specifically, if from tag-to-tag a report is resolved, then becomes outstanding again, our results table irrecoverably drops the information that the report was one point resolved. This will result in CodeChecker calculating the set of outstanding reports incorrectly. You can read more about this in [issue](https://github.com/Ericsson/codechecker/issues/3884) in section "Tag diffs -> Local-Remote -> the footnote with the double asterisk (**)".
```
Comment (dkrupp) [inline docs/web/diff.md:769]: I would rephrase this as follows:

Comparing past tags, may yield incorrect results in specific circumstances.
CodeChecker does not store a full state change history of a report, it applies a simplification for reopened reports.

It only stores the date when a report was first detected in a run and the date when it was closed. If the report is closed and opened several times, it will only remember the first detection date, and the last close date. So CodeChecker assumes the worst: it regards a report which is on/and/off as it was always there until it is finally fixed.

In practice this means, that when you compare the run state at two past tags (after a report detection, but before final fix), a reopened report will be considered as always outstanding after first detection and not appear as new report in the diff.
You can read more about this in ...

File: docs/web/diff.md [hunk 5]
```diff
@@ -0,0 +1,902 @@
+# Comparing analysis results (Diff)
+
+Compare the outstanding reports in a "before" (_baseline_) and an "after" (_new_ or newline) analysis sets, and display the differences in between them the in the same 
+format as `results`. A report is outstanding if all of the following is true:
+* its detection status is _new_, _reopened_ or _unresolved_,
```
Comment (dkrupp) [inline docs/web/diff.md:5]: A report is outstanding in a given time T if all the following is true:
- if it was detected before T
- if it was closed after T, or not closed at all
- a report can be closed if becomes undetected, considered false positive, intentional bug.

The server stores the past states of a reports with certain simplifications (see section **), so it can tell for any past time points whether the report was outstanding or closed.

The local report directory only stores a single present state.

In CodeChecker the most recent status of a report is also represented in the detection status:
-_new_: the report is new compared to the last analysis
-_reopened_: the report is new compared to the last analysis and it is reappearing
-_unresolved_: the report is still detected
-_resolved_: the report is not detected anymore

This is useful to quickly see if for example the analysis set of the "master" branch is updated, what are the new/unresolved reports we still have to work with.

**Review by dkrupp — CHANGES_REQUESTED** [review]
Please refer to this document from the usage guide.
Also please connect the sections to real PR analysis, branch analysis use-cases which should be introduced in the intro section. The user should understand how to use these concepts in practice.

**Review by dkrupp — APPROVED** [review]
With the condition of fixin my comments in diff.md in https://github.com/Ericsson/codechecker/pull/4006#pullrequestreview-1661327221

**Comment by Szelethus** [comment]
I cherry-picked the docs changes into #4006, so it no longer has any dependencies.

File: docs/web/diff.md (hunk 5 above)
Comment (Szelethus) [reply docs/web/diff.md]: I added the "outstanding report" definition when we diff tags/timestamps.

I added links to the detection status and review status docs.

For the rest, this is just not the place to describe what detection statuses are, but rather their own page. On the fact that we ignore detection statuses for tags/timestamps there already is paragraph.
